server/sonar-server/src/main/java/org/sonar/server/computation/step/SqaleMeasuresStep.java

```diff
@@ -0,0 +1,191 @@
+        long developmentCosts = computeDevelopmentCost(file);
+        path.current().add(developmentCosts);
+        computeAndSaveMeasures(file, path);
+      }
+    }
+
+    private void computeAndSaveMeasures(Component component, Path<DevelopmentCost> path) {
+      saveDevelopmentCostMeasure(component, path.current());
+
+      double density = computeDensity(component, path.current());
+      saveDebtRatioMeasure(component, density);
+      saveSqaleRatingMeasure(component, density);
+
+      increaseParentDevelopmentCost(path);
+    }
+
+    private void saveDevelopmentCostMeasure(Component component, DevelopmentCost developmentCost) {
+      // the value of this measure is stored as a string because it can exceed the size limit of number storage on some DB
+      measureRepository.add(component, developmentCostMetric, newMeasureBuilder().create(Long.toString(developmentCost.getValue())));
+    }
+
+    private double computeDensity(Component component, DevelopmentCost developmentCost) {
+      double debt = getLongValue(measureRepository.getRawMeasure(component, technicalDebtMetric));
+      if (Double.doubleToRawLongBits(developmentCost.getValue()) != 0L) {
+        return debt / (double) developmentCost.getValue();
+      }
+      return 0d;
+    }
+
+    private void saveDebtRatioMeasure(Component component, double density) {
+      measureRepository.add(component, debtRatioMetric, newMeasureBuilder().create(100.0 * density));
+    }
+
+    private void saveSqaleRatingMeasure(Component component, double density) {
+      SqaleRatingGrid ratingGrid = new SqaleRatingGrid(sqaleRatingSettings.getRatingGrid());
+      int rating = ratingGrid.getRatingForDensity(density);
+      String ratingLetter = toRatingLetter(rating);
+      measureRepository.add(component, sqaleRatingMetric, newMeasureBuilder().create(rating, ratingLetter));
+    }
+
+    private void increaseParentDevelopmentCost(Path<DevelopmentCost> path) {
+      if (!path.isRoot()) {
+        // increase parent's developmentCost with our own
+        path.parent().add(path.current().getValue());
+      }
+    }
+  }
+
+  private long computeDevelopmentCost(Component file) {
+    String languageKey = file.getFileAttributes().getLanguageKey();
+    String sizeMetricKey = sqaleRatingSettings.getSizeMetricKey(languageKey);
+    Metric sizeMetric = metricRepository.getByKey(sizeMetricKey);
+    return getLongValue(measureRepository.getRawMeasure(file, sizeMetric)) * sqaleRatingSettings.getDevCost(languageKey);
+  }
+
+  private static long getLongValue(Optional<Measure> measure) {
+    if (!measure.isPresent()) {
+      return 0L;
+    }
+    return getLongValue(measure.get());
+  }
+
+  private static long getLongValue(Measure measure) {
+    switch (measure.getValueType()) {
+      case INT:
+        return measure.getIntValue();
+      case LONG:
+        return measure.getLongValue();
+      case DOUBLE:
+        return new Double(measure.getDoubleValue()).longValue();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove the boxing to "Double". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2153)